check if cachefile dependencies can be fulfilled

vchuravy opened the pull request:
Fixes #21266. I was not able to reproduce the original issue,
because the dependency chain changed and `DataArrays` also
depends on `SpecialFunctions`.

We are writing transitive dependencies to the `.ji` file,
because at the moment of loading the `.ji` files there is no
difference between transitive dependencies or direct dependencies.
Just a list of preconditions, that need to be fullfilled.

Instead of changing that fairly complex piece of logic during
feature freeze I decided to try a less disruptive approach
and check in `stale_cachefile`, if all dependencies are
available.

Any suggestions for tests?

vchuravy committed:
check if cachefile dependencies can be fulfilled

vchuravy commented on base/loading.jl:
```diff
@@ -742,7 +743,17 @@ function stale_cachefile(modpath::String, cachefile::String)
+
+        # Check if transitive dependencies can be fullfilled
+        for mod in keys(required_modules)
+            isdefined(Main, mod) && continue
```
It might be better to check here for `:Core, :Base, ...`.

tkelman replied:
why would staleness depend on the in memory state of Main?

vchuravy replied:
I am checking if the module is already loaded here, since if it is `read_verify_mod_list` from `dump.c`, won't fail. Just added a commit to be more explicit (it safes a few calls to `find_in_node_path`)

vchuravy commented:
A different solutions that would probably work, but is imho tricky to implement: Ensure a consistent ordering in the module dependency list, so that direct dependencies are loaded first. But at the point where that list is written we have no information about direct or indirect dependencies.

tkelman commented:
for testing, we should be able to create a stack of 3 placeholder modules. Precompile the top one, then remove the bottom one and change the middle to not depend on it any more. That should reproduce the failure on master I believe.

vtjnash reviewed (approved):
I hate to slow it down this much (I've seen `find_in_node_path` can be a hot spot), since it'll probably cause a measurable performance regression in #15048). But this looks valid.

musm commented on base/loading.jl:
```diff
@@ -699,7 +699,16 @@ function parse_cache_header(f::IO)
     @assert totbytes == 4 "header of cache file appears to be corrupt"
-    return modules, files
+    # read the list of modules that are required to be present during loading
+    required_modules = Dict{Symbol, UInt64}()
```
nitpick :P but no space between {Symbol,UInt64}

KristofferC replied:
According to?

musm replied:
consistency with https://github.com/JuliaLang/julia/blob/70fdbe83a3022363fadbedef344b88dc3f05ea88/base/loading.jl#L683

and

https://github.com/JuliaLang/julia/blob/70fdbe83a3022363fadbedef344b88dc3f05ea88/base/loading.jl#L683

vchuravy commented:
Added a test. (Developed against master with `Base.require`)

tkelman commented on test/compile.jl:
```diff
@@ -325,6 +325,41 @@ try
+          """)
+    rm(FooBarT_file)
+    @test Base.stale_cachefile(FooBarT2_file, joinpath(dir2, "FooBarT2.ji"))
+    # Before #21492 this would fail: Base.require(:FooBarT2)
```
why don't we run it then?

vchuravy replied:
My logic was that if the line above fails it should also fail (and we don't quite have a `@test_nothrow`), but you are right it shouldn't hurt also doing it here.

tkelman replied:
we could test that it returns `=== nothing` if that's what it's supposed to do

vchuravy commented:
I addressed all comments, there was one intermediate travis failure that is backed here https://gist.github.com/vchuravy/aa4a715b99ef33435892bfd934bcc577, but it seems unrelated.

vchuravy commented:
with regard to performance (all tests done on a clean PKGDIR and after `Pkg.init(); Pkg.add("DataFrames")`)

##### This branch
```
# Precompilation run
JULIA_PKGDIR=${HOME}/jl21266 time ./julia -e "using DataFrames"
21.93user 5.97system 0:21.54elapsed 129%CPU (0avgtext+0avgdata 271452maxresident)k
0inputs+6000outputs (0major+160067minor)pagefaults 0swaps

# After precompilation
JULIA_PKGDIR=${HOME}/jl21266 time ./julia -e "using DataFrames"
1.47user 0.74system 0:01.42elapsed 155%CPU (0avgtext+0avgdata 211916maxresident)k
0inputs+0outputs (0major+11971minor)pagefaults 0swaps
```

##### Current master
```
JULIA_PKGDIR=${HOME}/jl21266 time ./julia -e "using DataFrames"
21.55user 6.02system 0:21.14elapsed 130%CPU (0avgtext+0avgdata 271000maxresident)k
0inputs+6016outputs (0major+160987minor)pagefaults 0swaps

 JULIA_PKGDIR=${HOME}/jl21266 time ./julia -e "using DataFrames"
1.49user 0.76system 0:01.48elapsed 151%CPU (0avgtext+0avgdata 214772maxresident)k
0inputs+0outputs (0major+12437minor)pagefaults 0swaps
```

@vtjnash is there a better way to check the performance impact? Right now it seems imperceptible, but it might be worse if the `JULIA_PKGDIR` is on a remote filesystem.

vtjnash commented:
Those timings look good.

bjarthur commented:
this PR breaks Gadfly.jl's ability to pre-compile:

```
julia> using Gadfly
INFO: Recompiling stale cache file /Users/arthurb/.julia/lib/v0.6/Gadfly.ji for module Gadfly.
WARNING: The call to compilecache failed to create a usable precompiled cache file for module Compose. Got:
WARNING: Module Iterators uuid did not match cache file.
WARNING: eval from module Main to Gadfly:    
Expr(:call, Expr(:., :Base, :include_from_node1)::Any, "/Users/arthurb/.julia/v0.6/Compose/src/Compose.jl")::Any
  ** incremental compilation may be broken for this module **

WARNING: imported binding for Iterators overwritten in module Main
```

vchuravy commented:
Thanks for letting me know I am looking into it. What version of Compose/Gadfly are you using?

bjarthur commented:
i'm on the master branch of each.

KristofferC commented:
I have also gotten the `WARNING: Module Iterators uuid did not match cache file.` recently and that was before this commit. Are you sure this is the commit that broke it?

bjarthur commented:
i've checkout out e0c497b22289487aff417380cdbdefa77255dd3b and everything works fine.

fredrikekre commented:
I got the same error just now, and as for `@bjarthur` everything works if i checkout e0c497b . I tried removing `~/.julia/lib/v0.6/` but I get the same failure.
Found this from `using PGFPlots` but (for me) it boils down to the following error message from `using AxisArrays`

```julia
julia> using AxisArrays
INFO: Recompiling stale cache file /home/fredrik/.julia/lib/v0.6/AxisArrays.ji for module AxisArrays.
WARNING: The call to compilecache failed to create a usable precompiled cache file for module AxisArrays. Got:
WARNING: Module Iterators uuid did not match cache file.
ERROR: LoadError: Declaring __precompile__(true) is only allowed in module files being imported.
Stacktrace:
 [1] __precompile__(::Bool) at ./loading.jl:318
 [2] __precompile__() at ./loading.jl:314
 [3] include_from_node1(::String) at ./loading.jl:539
 [4] eval(::Module, ::Any) at ./boot.jl:235
 [5] require(::Symbol) at ./loading.jl:453
while loading /home/fredrik/.julia/v0.6/AxisArrays/src/AxisArrays.jl, in expression starting on line 1
```

vchuravy commented:
@vtjnash I might need your help here

In a fresh Julia session
```
julia> Base.module_uuid(Main.Iterators)
0x000003c21dc6b0cb

julia> using Iterators
WARNING: imported binding for Iterators overwritten in module Main

julia> Base.module_uuid(Main.Iterators)
0x0000057e0e7d6f6b

julia> Base.parse_cache_header(joinpath(first(Base.LOAD_CACHE_PATH), "Compose.ji"))
...,  :Iterators=>0x0000057e0e7d6f6b, ...
```

But when we are restoring the module in `src/dump.c` `read_verify_mod_list`, we first check if the binding is already defined in `Main` and so we do not load the `Iterators` module provided by the package and then have an uuid conflict with `Base.iterators`. This is a issue independent of the PR above.

tkelman commented:
x-ref https://github.com/JuliaLang/julia/pull/19108, possibly relevant?

vtjnash commented on base/loading.jl:
```diff
@@ -742,7 +743,22 @@ function stale_cachefile(modpath::String, cachefile::String)
+            if mod == :Main || mod == :Core || mod == :Base
+                continue
+            # Module is already loaded
+            elseif isdefined(Main, mod)
```
needs a check for `isbindingresolved(Main, mod)` first, to avoid causing `isdefined` to resolve the binding
